Terminal: allow PowerShell 7.4+ constrained language audit mode to work with shell integrations

skbtwiz opened the pull request:
# Fixes https://github.com/microsoft/vscode/issues/283151

## Description
PowerShell 7.4 introduced **Constrained Language Audit Mode**, where `$ExecutionContext.SessionState.LanguageMode` reports `ConstrainedLanguage` but runtime behavior is equivalent to `FullLanguage` Mode. This is intended only for logging what _would_ be blocked were ConstrainedLanguage mode actually being enforced seriously.

The PowerShell VSCode shell integration guard currently aborts unless `LanguageMode == FullLanguage`. This logic predates the existence of audit-only constrained language mode and was correct for its time. Presently however, it falsely blocks the shell integration from working in audit scenarios, even though PowerShell execution is fully functional.

This change updates the guard to allow shell integration when:

- `LanguageMode == FullLanguage`, or
- `LanguageMode == ConstrainedLanguage` **and** system lockdown policy is `Audit`

Enforced constrained language mode continues to be blocked.

## Why this change is safe
- Audit CLM is explicitly non-restrictive by design
- Enforced CLM behavior remains unchanged
- Backward compatible with all older PowerShell versions
- No elevation or system configuration changes required
- Uses PowerShell’s official `SystemPolicy.GetSystemLockdownPolicy()` API

## How to test
**Audit CLM scenario**
1. On Windows 11, enable a WDAC / App Control **audit-only** policy
2. Install PowerShell 7.4+. If step 1 went correctly, pwsh should be in ConstrainedLanguage Audit Mode
3. Open VS Code
4. Open the integrated terminal using `pwsh`
5. Verify shell integration initializes correctly (prompt tracking, command boundaries)

**Enforced CLM scenario**
1. Enable an enforced WDAC / App Control policy
2. Open the integrated terminal using `pwsh`
3. Verify shell integration does **not** initialize

**Baseline**
- Windows PowerShell 5.1 (FullLanguage) continues to work unchanged because it is too old to understand Audit Mode.

## Notes
The original language mode guard was introduced in #158548 and was correct at the time. This change adapts that logic to newer PowerShell semantics introduced in 7.4 without weakening security guarantees.

<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->

skbtwiz committed:
Add Terminal:Shell-Integration check for Pwsh 7.4 ConstrainedLanguage Audit mode

Enhance shell integration checks for constrained language audit mode introduced in PowerShell 7.4 and later. This new mode is identical in behavior to FullLanguage mode and deserves to receive equal treatment.

skbtwiz committed:
Merge branch 'main' into skbtwiz/shell-integration-constrained-mode-audit

Tyriar reviewed (commented):
@andyleejordan could you review this when you get a chance? 🙏

skbtwiz committed:
Improve shell integration language mode checks

Made changes as per reviewer request

daxian-dbw commented on src/vs/workbench/contrib/terminal/common/scripts/shellIntegration.ps1:
```diff
@@ -10,7 +10,19 @@ if ((Test-Path variable:global:__VSCodeState) -and $null -ne $Global:__VSCodeSta
 # Disable shell integration when the language mode is restricted
 if ($ExecutionContext.SessionState.LanguageMode -ne "FullLanguage") {
-	return;
+	$LanguageMode = $ExecutionContext.SessionState.LanguageMode
```
nitpick: How aboue moving this line outside of the if block starting at line 12, and change line 12 to `if ($LanguageMode -ne "FullLanguage")`?

daxian-dbw replied:
That's fine. This is a minor comment that can be ignored. But be noted that the if statement doesn't create a new scope, so in case it's running in constrained audit mode, the variable will be visible in the script block.

daxian-dbw commented on src/vs/workbench/contrib/terminal/common/scripts/shellIntegration.ps1:
```diff
@@ -10,7 +10,19 @@ if ((Test-Path variable:global:__VSCodeState) -and $null -ne $Global:__VSCodeSta
+	try {
+		# Allow ConstrainedLanguage only when Windows lockdown policy is Audit (PS 7.4+).
+		$Lockdown = [System.Management.Automation.Security.SystemPolicy]::GetSystemLockdownPolicy()
+		if ($Lockdown -and $Lockdown -ne [System.Management.Automation.Security.SystemEnforcementMode]::Audit) {
```
```suggestion
		if ($Lockdown -and $Lockdown -ne "Audit") {
```
Let's use the string "Audit" here. The enum member "Audit" is not versions prior to 7.4, so in case user is using a old version, `[System.Management.Automation.Security.SystemEnforcementMode]::Audit` would fail.

skbtwiz committed:
Change Audit enum check to string check

This was done because the enum check only works on PS 7.4+ and compatibility with earlier versions must be maintained.

Co-authored-by: Dongbo Wang <dongbow@microsoft.com>

daxian-dbw reviewed (approved):
LGTM!

skbtwiz commented:
Hello @anthonykim1 ,

Please review this PR if and when you get the opportunity. Since this is my first major contribution to a large open source project, I apologise in advance if this message is out of place. This PR would significantly improve QoL on VS Code as well as its fork IDEs as currently the IDE's Rich Terminal is broken on all of them if system is on ConstrainedLanguage Audit mode, which is true for majority of the new dev machines we have received at our company that are Copilot+/Recall certified.

Once again, no urgency implied, just that my team is eager to see these fixes on their machines. Currently we are manually applying these fixes after each update.
